### PR TITLE
Incorporate internal changes and publish

### PR DESCRIPTION
**Pull request by rgrempel**
See #1

### CODE REVIEW

**Comment by rgrempel** [comment]
This is some work-in-progress so far ... it compiles, but isn't finished yet.

File: src/Restful/Endpoint.elm [hunk 1]
```diff
@@ -72,64 +100,279 @@ The basic idea is to include in this type all those things about an endpoint
...
+| `key` | Your ID type. We usually use some kind of `EntityId`. |
+| `value` | Your value type. |
+| `posted` | The type you would use in POST requests, when creating a new value. May be missing some information from `value` which the backend will supply. May be the same as `value` if POST isn't special. |
+
+The fields you need to fill in for an endpoint are as follows:
+
+| Field | Usage |
+| ----- | ----- |
+| `decodeKey` | Given the JSON the backend returns for each item, how can we decode your `key` type? <p>If you're using a kind of `EntityId`, for the `key`, then you can just supply `decodeEntityId`. |
+| `decodeMultiple` | For requests like `select`, which return multiple values, what do we need to decode to get a list we can apply `decodeKey` or `decodeVvalue` to? |
```
Comment (amitaibu) [inline src/Restful/Endpoint.elm]: decodeVvalue => decodeValue

File: src/Restful/Endpoint.elm [hunk 2]
```diff
@@ -72,64 +100,279 @@ The basic idea is to include in this type all those things about an endpoint
...
+| `decodeSingle` | For requests like `get`, which return a single value, what do we need to decode before we apply `decodeKey` or `decodeValue`? That is, how must we traverse the JSON to get to the value which `decodeKey` or `decodeValue` can decode? |
+| `decodeValue` | Given the JSON the backend returns for each item, how can we decode your `value` type? |
+| `encodeParams` | Takes your `params` type and converts it into something we can feed to `HttpBuilder.withQueryParams`. So, you get type-safety for the params! If you never take params, you can supply `always []`. |
+| `encodePostedValue` | An encoder for your `postedValue`, used when creating new values using POST requests. May be the same as `encodeValue` if POST isn't special in your case. |
+| `encodeValue` | An encoder for your `value` type (e.g. for PUT). |
+| `keyToUrlPart` | Given your `key`, what should we put after the `path` in a URL to refer to this entity (e.g. for PATCH, PUT or DELETE)? |
```
Comment (amitaibu) [inline src/Restful/Endpoint.elm]: Use case?

File: src/Restful/Endpoint.elm [hunk 3]
```diff
@@ -72,64 +100,279 @@ The basic idea is to include in this type all those things about an endpoint
...
+        , encodePostedValue : posted -> Value
+        , kv : KeyValue key value
+        , mapError : Error -> error
+        , path : String
+        , tokenStrategy : TokenStrategy
+        }
+
+
+{-| Various things which relate to the key and value. These all need to be changed
+together ... they could be split up a bit if Elm had Rank-N types ... in that case,
```
Comment (amitaibu) [inline src/Restful/Endpoint.elm]: I'll need some explanation here :)

File: src/Restful/Endpoint.elm [hunk 4]
```diff
@@ -72,64 +100,279 @@ The basic idea is to include in this type all those things about an endpoint
...
+
+  - Your endpoint doesn't use any URL params. But you can change that with ...
+
+  - You're not using a custom error type. But you can change that with ...
+
+  - An access token, if provided, will be sent as a URL param named "access_token".
+    But you can change that with ...
+
+-}
+drupalEndpoint : String -> Decoder value -> (value -> Value) -> EndPoint Error () (EntityId a) value value
```
Comment (amitaibu) [inline src/Restful/Endpoint.elm]: Maybe move this one to `Restful.Endpoint.Drupal`

File: src/Restful/Endpoint.elm [hunk 5]
```diff
@@ -72,64 +100,279 @@ The basic idea is to include in this type all those things about an endpoint
...
+    But you can change that with ...
+
+  - Single values are returned as a single-elmeent JSON array, inside a `data` field.
+    But you can change that with ...
+
+  - Your endpoint doesn't use any URL params. But you can change that with ...
+
+  - You're not using a custom error type. But you can change that with ...
+
+  - An access token, if provided, will be sent as a URL param named "access_token".
```
Comment (amitaibu) [inline src/Restful/Endpoint.elm]: In Drupal you won't need to change it, so you can probably remove this part of the comment

File: src/Restful/Endpoint.elm [hunk 6]
```diff
@@ -72,64 +100,279 @@ The basic idea is to include in this type all those things about an endpoint
...
+        , path = path
+        , tokenStrategy = TokenUrlParam "access_token"
+        }
+
+
+{-| We can use two strategies to send an `AccessToken` to the backend -- either an
+HTTP header (with a key and value), or a param for the URL (with key and value).
+Use `tokenHeader` or `tokenUrlParam` to construct.
+-}
+type TokenStrategy
```
Comment (amitaibu) [inline src/Restful/Endpoint.elm]: In fact, RESTful accepts both as-well :)

File: src/Restful/Endpoint.elm [hunk 7]
```diff
@@ -276,72 +486,35 @@ This function assumes that the backend will send the full value back. If it won'
...
--}
-{-
-   upsert : BackendUrl -> Maybe AccessToken -> EndPoint error params key value -> Entity key value -> (Result error (Entity key value) -> msg) -> Cmd msg
-   upsert backendUrl accessToken endpoint (key, value) tagger =
-       let
-           queryParams =
-               accessToken
-                   |> Maybe.Extra.toList
-                   |> List.map (\token -> ( "access_token", token ))
+patch_ : BackendUrl -> Maybe AccessToken -> EndPoint error params key value posted -> key -> Value -> (Result error () -> msg) -> Cmd msg
```
Comment (amitaibu) [inline src/Restful/Endpoint.elm]: Maybe `patchNoDecode` instead of the `_` suffix?

**Comment by rgrempel** [comment]
This is now covering as much ground as I think is sensible, for the moment.

So, the next task is to do PR's against the internal projects that use this, to see whether it actually feels nice to use the new API. (And, to test it ... testing it here would be nice too, but it's so `RequestBuilder`, `Task` and `Cmd` based that it would be a bit interesting).

There's a bit of a chicken-and-egg problem, in that I'd like to do PR's against the internal projects before I actually publish this. I suppose I can do that via copying or symlinks or something, and then finalize the PR once this is actually published.

**Comment by rgrempel** [comment]
I've started a couple of PR's on internal projects to test the API changes here to see how nice they feel. One is in progress, and there's another project I haven't started on yet.

So far, the results are good, with a couple of changes already incorporated here. Possible further changes include:

- Do something nice for endpoints that are read-only ... i.e. make specifying an encoder optional, and reflect that in the types.

- Make another version of `patch` which uses the full value ... or perhaps just the `create` value, rather than an arbitrary `Value`. Since sometimes we want to say PATCH to the backend even if we're sending all we've got.

Combining those two thoughts, I wonder whether instead of `value` and `created` we might actually want something like `readValue` and `writeValue`. That way, we could start the endpoint with a `writeValue` of `Never`, and change it only once we have an encoder. (Of course, the `readValue` and the `writeValue` could be the same).

In cases where we have a separate value type for POST (i.e. the `createValue`), I wonder whether it might make sense to use that type for all writes? The extra information in the read value is supplied by the backend, after all, and probably can't be changed anyway. So, in cases where the `readValue` and `writeValue` are different, we'd ask for an encode for the `writeValue` only. Though, in that case, we'd probably force the client to write a bit of code to convert from `readValue` to `writeValue`, which seems slightly awkward .... it seems nicer just to write an encode that picks out the relevant bits from `readValue`. (Which is why POST is necessarily special, since it's the case where we don't have a `readValue` from the backend yet).

Well, I'll think about that a bit.

File: src/Restful/Endpoint.elm (hunk 7 above)
Comment (rgrempel) [reply src/Restful/Endpoint.elm]: Come to think of it, with the new API, I can write a `withoutDecoder` function that can be used with any of the requests (when you want to ignore the success result). So, I won't need the specific variants for `put` and `patch`.

File: src/Restful/Endpoint.elm (hunk 3 above)
Comment (rgrempel) [reply src/Restful/Endpoint.elm]: I'm thinking of a blog post on "The State of Fancy Types in Elm", covering:

- typeclasses
- Rank-N types
- higher-kinded polymorphism

... discussing how they would help, and some workarounds for not having them.

File: src/Restful/Endpoint.elm (hunk 2 above)
Comment (rgrempel) [reply src/Restful/Endpoint.elm]: For `keyToUrlPart`? We need it to convert the key to the part of the URL that represents the ID. It's more or less what `untag` used to be, but with a more descriptive name. (Well, apparently not perfectly descriptive!)

**Comment by rgrempel** [comment]
Another note to self: consider whether `toCmd404` and `toTask404` would benefit from different names. Perhaps `toCmdMaybe` and `toTaskMaybe`, since they wrap the result in a `Maybe`?

**Comment by rgrempel** [comment]
Thinking about PUT, POST and PATCH and the various types some more, I've had the following bits of clarity.

- We pretty much always will want a decoder for the full `value` type -- I suppose you could imagine "write-only" endpoints, but that's probably rare.

- Normally, when we want to do a PUT or PATCH, we'll have the full `value`, because normally we would read the value from the backend before we change it! So, `put` and `patch` may as well take the `value` type as a parameter ... if fields should be left out when transmitting to the server, the encoder can just leave them out.

- For `patch`, I had been allowing any arbitrary JSON `Value` to be transmitted to the backend (since one PATCH might use different fields than another). It occurs to me now that this can be made more type-safe by requiring `value -> Value` instead of `Value`. That is, instead of requiring the JSON `Value` itself, require a function from the `value` to the JSON `Value`. That way, we tie the production of the JSON to the `value` ... it's what you would do anyway, but this way we enforce it and let the compiler catch certain kinds of errors.

- There may be an unusual use case for sending a PATCH request having just a `key`, without a `value`. We might know about one field we definitely want to update, without knowing anything else about the `value`. So, perhaps there is a purpose for a variant of `patch` that just accepts an arbitrary `Value`. (ideally, we'd actually have a data-level representation of the fields, but I think that's best left for version 2.0).

- There is probably no use case for sending a PUT request when we only have the key -- if it's an existing resource, we really ought to read it first before replacing the whole thing! And, I suppose, if there is a case I'm not thinking of, it could probably be handled by doing a PATCH instead, and sending the full value if we like.

So, I think the separate `created` type is only relevant to POST ... for the others, we either use the `value`, or in the a `patch` variant, an arbitrary `Value`.

The remaining issue is how to implement a `read-only` Endpoint (which is fairly common) at the type-level, and in a way that is nicer to create. The easy way would be to add yet another type variable ...
instead of:

    EndPoint error key value created params

you'd have something like:

    EndPoint error key readValue writeValue createValue params

Then, `drupalEndpoint` (and `endpoint`) would take just two parameters, rather than three ... something like this:

    drupalEndpoint : String -> Decoder readValue -> EndPoint Error (EntityId a) readValue Never Never p

I haven't tried it yet, but I think the `Never` in the type parameters would initially prevent such an endpoint from being used with `put`, `post` or `patch`, which is what we'd want ... it would start as a read-only endpoint at the type level. Then, we'd have a couple of functions to alter that:

    withEncoder : (writeValue -> Value) -> EndPoint e k rv wv cv p -> EndPoint e k rv writeValue writeValue p

    withCreateEncoder : (createValue -> Value) -> EndPoint e k rv wv cv p -> EndPoint e k rv wv createValue

(Note that `withEncoder` would set both the `writeValue` and the `createValue`, so `withCreateEncoder` would need to be applied afterwards if the `createValue` is different. I think this makes sense, since otherwise we'd often need to call both functions with the same encoder, which seems like overkill).

I **think** this would work nicely -- the only real down-side is the extra type parameter (we'd now have 6). Perhaps there is a clever way to cut down on the number of type parameters? It's probably impossible, because we really do want `readValue`, `writeValue` and `createValue` to be possibly different from each other. 

Though, I suppose `writeValue` is always either:

- the same as `readValue`; or
- `Never`

... and `createValue` is always either:

- the same as `writeValue`; or
- different.

Perhaps there is something to be made of that?

I suppose we don't really need a `writeValue` that can theoretically be anything, since it should either be `readValue` or impossible. We could use a "phantom type" instead ... in fact, that might reduce the number of type variables we need, in practice, due to the use of type aliases. 

Well, in fact type aliases could help even without phantom types ... consider this:

    type alias ReadOnlyEndpoint error key value params =
        Endpoint error key value Never Never params

    type alias ReadWriteEndpoint error key value params =
        Endpoint error key value value value params

    type alias ReadWriteCreateEndpoint error key value createValue params =
        Endpoint error key value value createValue params

Oh, that's much nicer! By using the aliases, we can avoid specifying more explicit type variables than we actually need, even though all of them ultimately exist. And we still get the type-safety for `put` etc., since the type-signature for `put` would require that the `readValue` and the `writeValue` be the same (so you couldn't `put` to a `ReadOnlyEndpoint`).

Anyway, the phantom-type approach would do that a little differently ... something like:

    type ReadOnly
        = ReadOnly

    type ReadWrite
         = ReadWrite

    type Endpoint writeable error key value createValue params
        = ...

    type alias ReadOnlyEndpoint error key value params
        = EndPoint ReadOnly error key value () params

    type alias ReadWriteEndpoint error key value params
        = EndPoint ReadWrite error key value value params

    type alias ReadWriteCreateEndpoint error key value createValue params
        = EndPoint ReadWrite error key value createValue params

Then, the functions which configure an `EndPoint` would need to return the appropriate "phantom" `writeable` type. (It's phantom in the sense that it's not actually controlling the type of any variable -- it's just a marker we can use to enforce certain things in type signatures).

It works out to roughly the same thing, but does avoid the `Never` type.

Anyway, one of those two approaches should work, and the use of the type aliases will mitigate the extra type variable required.

**Comment by amitaibu** [comment]
> So, I think the separate created type is only relevant to POST

Indeed.

> Anyway, the phantom-type approach would do that a little differently

Yes, I think it looks a little nicer, but I haven't dived deep -- just the asthetics of not using `Never`

One more thing to consider, which I just have a use case: In RESTful we all to have a GET with multiple IDs (e.g. `api/items/1,2,3`), which is similar to doing a `select` on api/items with [filtering](https://github.com/RESTful-Drupal/restful/blob/7.x-1.x/docs/api_drupal.md#filter).

In other words, I think `get` should return `List ( key, value )` instead of `value`

**Comment by rgrempel** [comment]
I think I'll probably do a separate `getMany`, which takes a `List key` and returns a `List value` ... it's probably worth having both (since getting a single ID is a pretty common use-case, so it's nice to keep a `get` specialized for it).

Speaking of filters, something I have in the back of my mind for version 2.0 is code to help construct filters ... basically, to help constructing a `params -> List (String, String)`, with some of the logic that the Drupal backend can handle (the ands and ors etc.). Also, it might be nice (in a future version) to allow for a representation of the fields at the data level (like Yesod does), so we wouldn't necessarily need a free-form `params` ... we'd actually know what fields are available for constructing filters (and for patch requests etc.). (And then in version 3.0 we could do some auto-discovery ...).

Anyway, back to the present!

**Comment by rgrempel** [comment]
Actually, I suppose `getMay` should probably return a `List (key, value)` rather than a `List value`, since when you have multiple values, it isn't quite as trivial to match the keys and the values ... if we return tuples, we don't have to worry about matching the order.

**Comment by rgrempel** [comment]
Just thinking out loud about the `BackendUrl` for a moment.

From one point of view, it would be natural to include the `BackendUrl` in our new `Backend` type, since it pertains to the backend. Then, we could avoid requiring it over and over again with each CRUD request.

What would be awkward about that, though, is that we can currently construct an entire `Backend` and `EndPoint` statically -- i.e. without any run-time information. However, the `BackendUrl` is typically run-time information in our apps, because we may be running in one environment or another, with distinct URLs for backends.

Now, you could imagine creating a structure like `(BackendUrl, EndPoint)`, and then variations on `get` etc. which taken those, without using `BackendUrl` as as separate parameter. But then you'd need a `let` in your `update` functions to produce the `(BackendUrl, EndPoint)` tuple for each `EndPoint`, which would be a real pain ... the cure would be worse than the disease.

What would be better is creating a version of the `get` etc. functions with the `BackendUrl` partially applied. In fact, we already use that idiom in a couple of places ... we have a `let` defining things like `getForBackend` where we just partially apply the `BackendUrl`.

But we can make that idiom nicer ... at the moment, we have to do that manually each time. Instead, we could define a record type with all the functions partially applied, and then a single `applyBackendUrl` function that produces that record. Something like:

```elm
type alias CrudOperations e k v c p =
    { select : EndPoint e k v c p -> p -> CrudRequest e (QueryResult k v p)
    , selectWithRange : EndPoint e k v c p -> p -> Offset -> Maybe Range -> CrudRequest e (QueryResult k v p)
    , get : EndPoint e k v c p -> k -> CrudRequest e v
    , ...
    }

applyBackendUrl : BackendUrl -> CrudOperations e k v c p
```

I.e. each of the CRUD functions, but just specifying the `BackendUrl` once.

Then, we can do this in an `update` function:

```elm
let
   crud = applyBackendUrl backendUrl

in
    ...
    crud.get endpoint ...
  
    ...
    crud.put otherEndpoint ...
```

Anyway, I'll give this a try at some point.

As a pattern, this is a little like the way that some ML languages deal with modules ... they treat the module as a first-class data type, to which you can supply parameters. So, here we'd be creating a kind of "module" to which the `BackendUrl` is supplied as a parameter.

**Comment by rgrempel** [comment]
We could also have a variant that also partially applies the `AccessToken`, i.e.

```elm
applyBackendUrlAndAccessToken : BackendUrl -> AccessToken -> CrudOperations e k v c p
```

**Comment by rgrempel** [comment]
@amitaibu I've got three PR's, which you should see referenced above, that implement this new API.

Could you take a look at this when you take a chance, and those PR's, to see what you think? It's publishable now, I think.
